hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/event/BusProtoStuffMessageConverter.java

```diff
@@ -52,57 +59,96 @@ protected boolean supports(final Class<?> aClass) {
+        if (objectPayload instanceof byte[]) {
+            final byte[] payload = (byte[]) objectPayload;
+            final byte[] clazzHeader = new byte[EVENT_TYPE_LENGTH];
+            final byte[] content = new byte[payload.length - EVENT_TYPE_LENGTH];
 
-        try {
-            final Class<?> deserializeClass = ClassUtils
-                    .getClass(message.getHeaders().get(DEFAULT_CLASS_FIELD_NAME).toString());
-            if (payload instanceof byte[]) {
-                @SuppressWarnings("unchecked")
-                final Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(deserializeClass);
-                final Object deserializeEvent = schema.newMessage();
-                ProtobufIOUtil.mergeFrom((byte[]) message.getPayload(), deserializeEvent, schema);
-                return deserializeEvent;
-            }
-        } catch (final ClassNotFoundException e) {
-            LOG.error("Protostuff cannot find derserialize class", e);
-            throw new MessageConversionException(message, "Failed to read payload", e);
-        }
+            splitClazzHeaderAndContent(payload, clazzHeader, content);
 
+            final EventType eventType = readClassHeader(clazzHeader);
+            return readContent(eventType, content);
+        }
         return null;
     }
 
     @Override
     protected Object convertToInternal(final Object payload, final MessageHeaders headers,
             final Object conversionHint) {
-        checkIfHeaderMutable(headers);
+
+        final byte[] clazzHeader = writeClassHeader(payload.getClass());
+
+        final byte[] writeContent = writeContent(payload);
+
+        final byte[] body = new byte[clazzHeader.length + writeContent.length];
+
+        mergeClassHeaderAndContent(clazzHeader, writeContent, body);
+
+        return body;
+    }
+
+    private static Object readContent(final EventType eventType, final byte[] content) {
+        final Class<?> targetClass = eventType.getTargetClass();
+        if (targetClass == null) {
+            LOG.error("Cannot read clazz header for given EventType value {}, missing mapping", eventType.getValue());
+            throw new MessageConversionException("Missing mapping of EventType for value " + eventType.getValue());
+        }
+        @SuppressWarnings("unchecked")
+        final Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(targetClass);
+        final Object deserializeEvent = schema.newMessage();
+        ProtobufIOUtil.mergeFrom(content, deserializeEvent, schema);
+        return deserializeEvent;
+    }
+
+    private static void mergeClassHeaderAndContent(final byte[] clazzHeader, final byte[] writeContent,
+            final byte[] body) {
+        System.arraycopy(clazzHeader, 0, body, 0, clazzHeader.length);
+        System.arraycopy(writeContent, 0, body, clazzHeader.length, writeContent.length);
+    }
+
+    private static void splitClazzHeaderAndContent(final byte[] payload, final byte[] clazzHeader,
```

Comment:
Same here with `content`. Rename the method to e.g. `extractContent(...)` and return the content instead of passing a empty array into this method.